### DISCUSSION TITLE
Mobile App

### DISCUSSION THREAD

**Discussion by macuno7** [FiloSottile/age#365] [General]
Are there any plans to bring age to mobile platforms?

With a increasing number of people who are primarily using mobile phones to communicate and exchange data, a mobile implementation (Android/IOS) of age seems like a good idea. 

Is anyone possibly already working on such an implementation?

What are the thoughts of the developers on this idea?

Are there any expected issues with bringing age to mobile devices?

**Comment by str4d** [comment]
There are no issues from a technical perspective. I assume Go can be used in Android and iOS apps, and I know that Rust can, so the existing `age` Go library or `age` Rust crate would work perfectly fine as a backend, and the actual GUI frontend could be implemented entirely separately.

The main question is actually a UI / UX one: what should file encryption / decryption look like on mobile? It probably makes sense for some kind of standalone GUI to exist, but it could also be a webapp like my (beta, still buggy) https://rage-encry.pt or be integrated into other apps.

**Comment by str4d** [comment]
On a related note, we also need to figure out how age plugins might work for non-desktop situations (like webapps and mobile apps) where there isn't a `PATH` to rely on. For plugins written in the same language as the age library being used, they could just be bundled into the GUI, but in every other case it gets complicated. Let's leave this part to a separate discussion.

**Comment by khekmatyar** [comment]
There is a [repository](https://github.com/khekmatyar/SharedSecret) for a mobile app frontend in case anyone here is interested to help out with that. 
Any feedback on the described approach would be most welcome!

**Comment by msfjarvis** [comment]
I'm one of the maintainers of [Android Password Store](https://github.com/Android-Password-Store/Android-Password-Store) and we've successfully built the `age` library for mobile and confirmed that it indeed does the job. We're currently in the process of writing an `age` implementation in Kotlin to support it as an additional cryptographic backend alongside the existing GPG one much like how [passage](https://github.com/FiloSottile/passage) ports [pass](https://passwordstore.org/) to age, but it's a stretch goal that is not expected to be delivered within the first half of 2022 due to other priorities.

**Reply by subnut** [reply]
Is there any work being done currently on this?

How did you manage to build `age` for android? A few instructions would be helpful.

**Reply by msfjarvis** [reply]
There is not.

You can trivially build age with gomobile, in our experiments we implemented a basic shim library that used age as a dependency. That code can be found [here](https://msfjarvis.dev/apsg/age-mobile).

**Reply by clach04** [reply]
@msfjarvis possibly dumb question, is Java an option for you? https://github.com/exceptionfactory/jagged looks like it may be an alternative without native compilation. That should work for Android but if you have other targets, it may not be an option

**Reply by msfjarvis** [reply]
> @msfjarvis possibly dumb question, is Java an option for you? https://github.com/exceptionfactory/jagged looks like it may be an alternative without native compilation. That should work for Android but if you have other targets, it may not be an option

We did go ahead and write a JVM-native implementation without FFI at a later date: https://github.com/android-password-store/kage
